drivers/sensor/bmm150/bmm150.c

```diff
@@ -0,0 +1,595 @@
+	[REGULAR_PRESET] = {9, 15, 10},
+	[ENHANCED_REGULAR_PRESET] = {15, 27, 10},
+	[HIGH_ACCURACY_PRESET] = {47, 83, 20}
+};
+
+static int bmm150_set_power_mode(struct device *dev,
+				enum bmm150_power_modes mode,
+				int state)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+
+	switch(mode){
+		case BMM150_POWER_MODE_SUSPEND:
+			if(i2c_reg_update_byte(data->i2c,
+						config->i2c_slave_addr,
+						BMM150_REG_POWER,
+						BMM150_MASK_POWER_CTL,
+						!state) < 0) {
+				return -EIO;
+			}
+			k_busy_wait(5 * USEC_PER_MSEC);
+			
+			return 0;
+		case BMM150_POWER_MODE_SLEEP:
+			return i2c_reg_update_byte(data->i2c,
+						   config->i2c_slave_addr,
+						   BMM150_REG_OPMODE_ODR,
+						   BMM150_MASK_OPMODE,
+						   BMM150_MODE_SLEEP <<
+						   BMM150_SHIFT_OPMODE);
+			break;
+		case BMM150_POWER_MODE_NORMAL:
+			return i2c_reg_update_byte(data->i2c,
+						   config->i2c_slave_addr,
+						   BMM150_REG_OPMODE_ODR,
+						   BMM150_MASK_OPMODE,
+						   BMM150_MODE_NORMAL <<
+						   BMM150_SHIFT_OPMODE);
+			break;
+	}
+	
+	return -ENOTSUP; 
+
+}
+
+static int bmm150_set_odr(struct device *dev, u8_t val)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+	u8_t i;
+
+	for(i=0; i < ARRAY_SIZE(bmm150_samp_freq_table); ++i) {
+		if (val <= bmm150_samp_freq_table[i].freq) {
+			return i2c_reg_update_byte(data->i2c,
+						   config->i2c_slave_addr,
+						   BMM150_REG_OPMODE_ODR,
+						   BMM150_MASK_ODR,
+						   (bmm150_samp_freq_table[i].
+						   reg_val <<
+						   BMM150_SHIFT_ODR));
+		}
+	}
+	return -ENOTSUP;
+}
+
+#if defined(BMM150_SET_ATTR)
+static int bmm150_read_rep_xy(struct device *dev)
+{
+	struct bmm150_data *data = dev->driver->data;
+	const struct bmm150_config *config = dev->config->config_info;
+	int reg_val;
+
+	if(i2c_reg_read_byte(data->i2c, config->i2c_slave_addr,
+				BMM150_REG_REP_XY, &reg_val) < 0) {
+		return -EIO;
+	}
+
+	data->rep_xy = BMM150_REGVAL_TO_REPXY((int)(reg_val));
+
+	return 0;
+}
+
+static int bmm150_read_rep_z(struct device *dev)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+	u8_t reg_val;
+
+	if (i2c_reg_read_byte(data->i2c, config->i2c_slave_addr,
+				BMM150_REG_REP_Z, &reg_val) < 0) {
+		return -EIO;
+	}
+
+	data->rep_z = BMM150_REGVAL_TO_REPZ((int)(reg_val));
+
+	return 0;
+}
+
+static int bmm150_compute_max_odr(struct device *dev, int rep_xy,
+					  int rep_z, int *max_odr)
+{
+	struct bmm150_data *data = dev->driver_data;
+
+	if(rep_xy == 0) {
+		if (data->rep_xy <=0) {
+			if (bmm150_read_rep_xy(dev) <0 ) {
+				return -EIO;
+			}
+		}
+		rep_xy = data->rep_xy;
+	}
+
+	if(rep_z == 0) {
+		if (data->rep_z <=0) {
+			if (bmm150_read_rep_z(dev) <0 ) {
+				return -EIO;
+			}
+		}
+		rep_z = data->rep_z;
+	}
+
+	*max_odr = 1000000 / (145 * rep_xy + 500 * rep_z + 980);
```

Comment:
Please add a comment in the code that says this equation comes directly from the datasheet